### PR TITLE
8316017: Refactor timeout handler in PassFailJFrame

### PR DESCRIPTION
**Pull request by aivanov-jdk**
Refactored timeout handling in PassFailJFrame:

1. Managing the timer and formatting the time left is inside `TimeoutHandler` class.
2. The class handles timer events and updates the label accordingly.

This is implemented on top of #15665.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316017](https://bugs.openjdk.org/browse/JDK-8316017): Refactor timeout handler in PassFailJFrame (**Bug** - P4)

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [39ebc7c8](https://git.openjdk.org/jdk/pull/15668/files/39ebc7c8a92bf42f5fac52c35ea5ddf72e4aa2da)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15668/head:pull/15668` \
`$ git checkout pull/15668`

Update a local copy of the PR: \
`$ git checkout pull/15668` \
`$ git pull https://git.openjdk.org/jdk.git pull/15668/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15668`

View PR using the GUI difftool: \
`$ git pr show -t 15668`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15668.diff">https://git.openjdk.org/jdk/pull/15668.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15668#issuecomment-1714377378)

### CODE REVIEW

**Comment by lawrence-andrew** [comment]
Looks good to me.

File: test/jdk/java/awt/regtesthelpers/PassFailJFrame.java [hunk 1]
```diff
@@ -314,6 +303,54 @@ public interface WindowCreator {
...
+            updateTime(testTimeOut);
+        }
+
+        @Override
+        public void actionPerformed(ActionEvent e) {
+            long leftTime = endTime - System.currentTimeMillis();
+            if ((leftTime < 0) || failed) {
+                timer.stop();
+                testFailedReason = FAILURE_REASON
+                                   + "Timeout User did not perform testing.";
```
Comment (prrace) [inline test/jdk/java/awt/regtesthelpers/PassFailJFrame.java:329]: If the test was explicitly failed by the user (ie failed==true), why does the reason say that there was a time out ?

Comment (aivanov-jdk) [reply test/jdk/java/awt/regtesthelpers/PassFailJFrame.java:329]: Huh, it's a good question. I didn't notice it. It's the original code that I moved into a dedicated class.

It shouldn't change the failure reason if it's already set.

The timer isn't stopped in handlers for the failure, so it's possible that timer changes it. We should also stop the timer explicitly when failure or success is handled.

Comment (aivanov-jdk) [reply test/jdk/java/awt/regtesthelpers/PassFailJFrame.java]: @prrace Eventually, I refactored handling of `failureReason`.  I removed `failed` and `timeout` fields; the `failureReason` is set via a synchronised method to ensure there cannot be data race. If `failureReason` is not `null`, the test fails. After `failureReason` is set to a non-`null` value, it cannot be changed.

The timeout handler relies on this behaviour: it calls `setFailureReason` unconditionally; if `failureReason` is already set, the timeout reason is ignored.

**Review by prrace — APPROVED** [review]

**Pull request closed without merging**